core/modules/qdisp/Executive.h

```diff
@@ -155,31 +147,46 @@ class Executive {
     Config _config; ///< Personal copy of config
     util::Flag<bool> _empty;
     std::shared_ptr<MessageStore> _messageStore; ///< MessageStore for logging
-    XrdSsiService* _service; ///< RPC interface
-    RequesterMap _requesters; ///< RequesterMap for results from submitted tasks
-    JobStatusPtrMap _statuses; ///< Statuses of submitted tasks
+    XrdSsiService* _xrdSsiService; ///< RPC interface
+    JobMap _jobMap; ///< Contains information about all jobs.
+    JobMap _incompleteJobs; ///< Map of incomplete jobs.
 
     /** Execution errors */
     util::MultiError _multiError;
 
     int _requestCount; ///< Count of submitted tasks
-    bool _cancelled; ///< Has execution been cancelled?
+    util::Flag<bool> _cancelled; ///< Has execution been cancelled?
 
     // Mutexes
-    std::mutex _requestersMutex;
+    std::mutex _incompleteJobsMutex; ///< protect incompleteJobs map.
 
     /** Used to record execution errors */
     mutable std::mutex _errorsMutex;
 
-    std::condition_variable _requestersEmpty;
-    mutable std::mutex _statusesMutex;
-    std::mutex _retryMutex;
-    std::mutex _cancelledMutex;
+    std::condition_variable _allJobsComplete;
+    mutable std::mutex _jobsMutex;
 
     typedef std::map<int,int> IntIntMap;
     IntIntMap _retryMap; ///< Counter for task retries.
 
-}; // class Executive
+};
+
+class MarkCompleteFunc {
+public:
+    typedef std::shared_ptr<MarkCompleteFunc> Ptr;
+
+    MarkCompleteFunc(Executive* e, int jobId) : _executive(e), _jobId(jobId) {}
+
+    virtual void operator()(bool success) {
```

Comment:
Is it polymorphic, do you need `virtual` here? If yes then destructor should be virtual too.